dask_learn/solvers.py

```diff
@@ -0,0 +1,255 @@
+
+    classes = kwargs.pop('classes', False) or da.unique(y)
+    if len(classes) != 2:
+        raise ValueError("gradient solver doesn't support multiclass")
+    coded_y = (y == classes[1]).astype('i1')
+
+    w, c = gradient_descent(X, coded_y, fit_intercept=self.fit_intercept,
+                            max_iter=self.max_iter, tol=self.tol,
+                            verbose=self.verbose)
+    self.coef_ = w
+    self.intercept_ = c
+    self.classes_ = classes
+
+
+def _intercept_dot(X, w, c):
+    return X.dot(w) + c
+
+
+def _intercept_norm(w, c):
+    return ((w**2).sum() + c**2)**0.5
+
+
+def gradient_descent(X, y, fit_intercept=True, max_iter=100, tol=1e-4,
+                     verbose=False):
+    """Solve a logistic regression problem using gradient descent"""
+
+    log = print if verbose else lambda x: x
+    # Tuning Params
+    first_backtrack_mult = 0.1
+    next_backtrack_mult = 0.5
+    armijo_mult = 0.1
+    step_growth = 1.25
+    step_size = 1.0
+
+    # Init
+    w = np.zeros(X.shape[1])
+    c = 0.0
+    backtrack_mult = first_backtrack_mult
+
+    log('##       -f        |df/f|    |dw/w|    step\n'
+        '-------------------------------------------')
+    for k in range(1, max_iter + 1):
+        # Compute the gradient
+        Xw = _intercept_dot(X, w, c)
+        eXw = da.exp(Xw)
+        f = da.log1p(eXw).sum() - y.dot(Xw)
+        mult = eXw/(eXw + 1) - y
+        grad = X.T.dot(mult)
+        c_grad = mult.sum() if fit_intercept else 0.0
+        Xgrad = _intercept_dot(X, grad, c_grad)
+
+        Xw, f, grad, Xgrad, c_grad = da.compute(Xw, f, grad, Xgrad, c_grad)
+
+        step_len = _intercept_norm(grad, c_grad)
+
+        # Compute the step size using line search
+        old_w = w
+        old_c = c
+        old_Xw = Xw
+        old_f = f
+        for ii in range(100):
+            w = old_w - step_size * grad
+            if fit_intercept:
+                c = old_c - step_size * c_grad
+            if ii and np.array_equal(w, old_w) and c == old_c:
+                step_size = 0
+                break
+            Xw = old_Xw - step_size * Xgrad
+            # This prevents overflow
+            if np.all(Xw < 700):
+                eXw = np.exp(Xw)
+                f = np.log1p(eXw).sum() - np.dot(y, Xw)
+                df = old_f - f
+                if df >= armijo_mult * step_size * step_len**2:
+                    break
+            step_size *= backtrack_mult
+
+        if step_size == 0:
+            log('No more progress')
+            break
+        df /= max(f, old_f)
+        dw = (step_size * step_len /
+              (_intercept_norm(w, c) + step_size * step_len))
+        log('%2d  %.6e %-.2e  %.2e  %.1e' % (k, f, df, dw, step_size))
+        if df < tol:
+            log('Converged')
+            break
+        step_size *= step_growth
+        backtrack_mult = next_backtrack_mult
+    else:
+        log('Maximum Iterations')
+    return w[None, :], np.atleast_1d(c)
```

Comment:
Instead of re-implementing yet-another gradient descent solver, why not re-using the very efficient L-BFGS solver of scipy and use dask to parallelize the gradient computation (`fprime`)?

http://docs.scipy.org/doc/scipy-0.17.1/reference/generated/scipy.optimize.fmin_l_bfgs_b.html